File: examples/dummyTest.py
#!/usr/bin/env python

#
# Copyright 2014-2016 Sandia Corporation. Under the terms of Contract
# DE-AC04-94AL85000, there is a non-exclusive license for use of this work
# by or on behalf of the U.S. Government. Export of this program may require
# a license from the United States Government.
#
# This file is part of the Power API Prototype software package. For license
# information, see the LICENSE file in the top level directory of the
# distribution.
#

import pwr as pow 
import time

cntxt = pow.CntxtInit( pow.CNTXT_DEFAULT, pow.ROLE_APP, "App" )
assert( cntxt )

self = pow.CntxtGetEntryPoint( cntxt )
assert( self )

objName = pow.ObjGetTypeString( pow.ObjGetType( self ) )

print "I'm a {0}".format( objName )

parent = pow.ObjGetParent( self )
assert( 0 != parent  )

children = pow.ObjGetChildren( self );
assert( children )

for i in range(0,pow.GrpGetNumObjs(children)): 
	print "child {0}".format( pow.ObjGetName( pow.GrpGetObjByIndx( children, i ) ) ) 

retval,value,ts = pow.ObjAttrGetValue( self, pow.ATTR_VOLTAGE )
assert( retval == pow.RET_INVALID )

retval,value,ts = pow.ObjAttrGetValue( self, pow.ATTR_POWER )
assert( retval == pow.RET_SUCCESS )

retval, timevalue = pow.TimeConvert( ts )

print "ObjAttrGetValue(ATTR_POWER) value={0} ts={1}".format(value,time.ctime(timevalue))

value = 25.812;
print "ObjAttrSetValue(ATTR_POWER) value={0}".format(value)
retval = pow.ObjAttrSetValue( self, pow.ATTR_POWER, value );
assert( retval == pow.RET_SUCCESS );

name = [pow.ATTR_POWER]

status = pow.StatusCreate();

retval, value, ts = pow.ObjAttrGetValues( self, name, status )
assert( retval == pow.RET_SUCCESS )

retval, timevalue = pow.TimeConvert( ts );

print "ObjAttrGetValues(ATTR_POWER) value={0} ts={1}".format(value,time.ctime(timevalue))

value = [100.10]
print "ObjAttrSetValues(ATTR_POWER) value={0}".format(value)
retval = pow.ObjAttrSetValues( self, name, value, status )
assert( retval == pow.RET_SUCCESS );

retval, value, ts = pow.ObjAttrGetValue( self, pow.ATTR_POWER )
assert( retval == pow.RET_SUCCESS )

assert( value == 200.20 );

retval, timevalue = pow.TimeConvert( ts );
print "ObjAttrGetValue(ATTR_POWER) value={0} ts={1}".format(value,time.ctime(timevalue));

grp = pow.CntxtGetGrpByType( cntxt, pow.OBJ_CORE )
assert( grp );
assert( pow.GrpGetNumObjs( grp ) )

value = 0.1
print "GrpAttrSetValue(ATTR_POWER) value={0}\n".format(value)
retval = pow.GrpAttrSetValue( grp, pow.ATTR_POWER, value, status )
assert( retval == pow.RET_SUCCESS )

retval, value, ts = pow.ObjAttrGetValue( self, pow.ATTR_POWER )
assert( retval == pow.RET_SUCCESS )
assert( value == 0.2 );

retval, timevalue = pow.TimeConvert( ts )
print "ObjAttrGetValue(ATTR_POWER) value={0} ts={1}".format( value,time.ctime(timevalue))

core = pow.GrpGetObjByIndx( grp, 0 );
assert( core )

coreStat = pow.ObjCreateStat( core, pow.ATTR_POWER, pow.ATTR_STAT_AVG )
assert( coreStat )

retval = pow.StatStart( coreStat )
assert( retval == pow.RET_SUCCESS )

time.sleep( 0.00001 ) 

retval, value, statTimes = pow.StatGetValue( coreStat )
assert( retval == pow.RET_SUCCESS )

print "StatGetValue(ATTR_POWER) value={0}".format( value )
print "StatGetValue(ATTR_POWER) start={0}".format( statTimes.start )

print "StatGetValue(ATTR_POWER) stop={0}".format( statTimes.stop )

if ( statTimes.instant != pow.TIME_NOT_SET ):
    print "StatGetValue(ATTR_POWER) instant={0}".format( statTimes.instant )

pow.StatDestroy( coreStat )

pow.CntxtDestroy( cntxt )

File: examples/dummyTest.c
/* 
 * Copyright 2014-2016 Sandia Corporation. Under the terms of Contract
 * DE-AC04-94AL85000, there is a non-exclusive license for use of this work 
 * by or on behalf of the U.S. Government. Export of this program may require
 * a license from the United States Government.
 *
 * This file is part of the Power API Prototype software package. For license
 * information, see the LICENSE file in the top level directory of the
 * distribution.
*/

#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <string.h>
#include <unistd.h>

#include <pwr.h> 

char* myctime(const time_t *timep);

int main( int argc, char* argv[] )
{
    PWR_Grp     grp;
    PWR_Obj     self;
    PWR_Cntxt   cntxt;
    time_t      time;
    int         rc;
    double       value;
    PWR_Time ts;
	PWR_Status  status;

    // Get a context
    rc = PWR_CntxtInit( PWR_CNTXT_DEFAULT, PWR_ROLE_APP, "App", &cntxt );
    assert( PWR_RET_SUCCESS == rc );

    rc = PWR_CntxtGetEntryPoint( cntxt, &self );
    assert( PWR_RET_SUCCESS == rc );
    
	PWR_ObjType objType;
	PWR_ObjGetType( self, &objType );
    printf("I am a `%s`\n", PWR_ObjGetTypeString( objType ) ); 

    PWR_Obj parent;
	rc = PWR_ObjGetParent( self, &parent );
	assert( rc >= PWR_RET_SUCCESS );	

    PWR_Grp children;
	rc = PWR_ObjGetChildren( self, &children );
	assert( rc >= PWR_RET_SUCCESS );	

    int i;
    for ( i = 0; i < PWR_GrpGetNumObjs(children); i++ ) {
		char name[100];
		PWR_Obj obj;
        PWR_GrpGetObjByIndx( children, i, &obj );
		PWR_ObjGetName( obj, name, 100 );
 
        printf("child %s\n", name );
    }

    rc = PWR_ObjAttrGetValue( self, PWR_ATTR_VOLTAGE, &value, &ts );
    assert( PWR_RET_INVALID == rc );

    rc = PWR_ObjAttrGetValue( self, PWR_ATTR_ENERGY, &value, &ts );
    assert( PWR_RET_SUCCESS == rc );

    PWR_TimeConvert( ts, &time );
    printf("PWR_ObjAttrGetValue(PWR_ATTR_ENERGY) value=%f ts=`%s`\n",
													value,myctime(&time));
    
    value = 25.812;
    printf("PWR_ObjAttrSetValue(PWR_ATTR_ENERGY) value=%f\n",value);
    rc = PWR_ObjAttrSetValue( self, PWR_ATTR_ENERGY, &value );
    assert( PWR_RET_SUCCESS == rc );

    PWR_AttrName name = PWR_ATTR_ENERGY;
     
	rc =  PWR_StatusCreate( cntxt, &status );
	assert( PWR_RET_SUCCESS == rc );

    rc = PWR_ObjAttrGetValues( self, 1, &name, &value, &ts, status );  
    assert( PWR_RET_SUCCESS == rc );

    PWR_TimeConvert( ts, &time );
    printf("PWR_ObjAttrGetValues(PWR_ATTR_ENERGY) value=%f ts=`%s`\n", 
													value, myctime( &time ) );

    value = 100.10;
    printf("PWR_ObjAttrSetValues(PWR_ATTR_ENERGY) value=%f\n",value);
    rc = PWR_ObjAttrSetValues( self, 1, &name, &value, status );  
    assert( PWR_RET_SUCCESS == rc );

    rc = PWR_ObjAttrGetValue( self, PWR_ATTR_ENERGY, &value, &ts );
    assert( PWR_RET_SUCCESS == rc );


    PWR_TimeConvert( ts, &time );
    printf("PWR_ObjAttrGetValue(PWR_ATTR_ENERGY) value=%f ts=`%s`\n",
													value,myctime(&time));

    rc = PWR_CntxtGetGrpByType( cntxt, PWR_OBJ_CORE, &grp );
    assert ( PWR_RET_SUCCESS == rc );

    value = 0.1;
    printf("PWR_GrpAttrSetValue(PWR_ATTR_ENERGY) value=%f\n", value);
    rc = PWR_GrpAttrSetValue( grp, PWR_ATTR_ENERGY, &value, status );
    assert( PWR_RET_SUCCESS == rc );

    rc = PWR_ObjAttrGetValue( self, PWR_ATTR_ENERGY, &value, &ts );
    assert( PWR_RET_SUCCESS == rc );

    PWR_TimeConvert( ts, &time );
    printf("PWR_ObjAttrGetValue(PWR_ATTR_ENERGY) value=%f ts=`%s`\n",
													value,myctime(&time));

    PWR_Obj node;
    rc = PWR_GrpGetObjByIndx( grp, 0, &node );
    assert( PWR_RET_SUCCESS == rc );

	PWR_Stat nodeStat;
	rc = PWR_ObjCreateStat( node, PWR_ATTR_ENERGY, 
						PWR_ATTR_STAT_AVG, &nodeStat );
	assert( PWR_RET_SUCCESS == rc );

	rc = PWR_StatStart( nodeStat );
    assert( PWR_RET_SUCCESS == rc );

	sleep(1);
	PWR_TimePeriod statTimes;
    statTimes.start = statTimes.stop = PWR_TIME_UNINIT;

	rc = PWR_StatGetValue( nodeStat, &value, &statTimes );
    assert( PWR_RET_SUCCESS == rc );

    printf("PWR_StatGetValue(PWR_ATTR_ENERGY) value=%lf\n", value );
    printf("PWR_StatGetValue(PWR_ATTR_ENERGY) start=%lf\n", 
                                    (double)statTimes.start / 1000000000 );

   	printf("PWR_StatGetValue(PWR_ATTR_ENERGY) stop=%lf\n",
                                    (double) statTimes.stop / 1000000000);

	if ( statTimes.instant != PWR_TIME_UNINIT ) {
    	printf("PWR_StatGetValue(PWR_ATTR_ENERGY) instant=%lf\n",
									(double)statTimes.instant / 1000000000 );
	}
	PWR_StatDestroy( nodeStat );

	PWR_Stat nodesStat;
	rc = PWR_GrpCreateStat( grp, PWR_ATTR_ENERGY, PWR_ATTR_STAT_AVG, &nodesStat);
	assert( PWR_RET_SUCCESS == rc );

    int grpSize = PWR_GrpGetNumObjs( grp ); 
	PWR_TimePeriod* statTimes2 = malloc( grpSize * sizeof(*statTimes2) );
	double* values = malloc( grpSize * sizeof(*values) );

	for ( i = 0; i < grpSize; i++ ) {
        statTimes2[i].start = statTimes2[i].stop = PWR_TIME_UNINIT;
    } 
	rc = PWR_StatStart( nodesStat );
	assert( PWR_RET_SUCCESS == rc);

	sleep(1);

	rc = PWR_StatGetValues( nodesStat, &values[0], &statTimes2[0]);
	assert( PWR_RET_SUCCESS == rc );

	PWR_StatDestroy( nodesStat );

	for ( i = 0; i < grpSize; i++ ) {
		printf("stat: value=%lf start=%lf stop=%lf\n", values[i],
            (double) statTimes2[i].start/1000000000, 
            (double)statTimes2[i].stop/1000000000 );
	}

	//PWR_CntxtDestroy( cntxt );
    return 0;
}
char* myctime(const time_t *timep)
{
    char* tmp = ctime(timep);
    tmp[ strlen(tmp) - 1 ] = 0; 
    return tmp;
}
